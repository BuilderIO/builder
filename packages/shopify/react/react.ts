export { AssignBlock } from './components/Assign';
export { CaptureBlock } from './components/Capture';
export { ConditionBlock } from './components/Condition';
export { UnlessBlock } from './components/Unless';
export { ProductBox } from './components/ProductBox';
export { ShopifySection } from './components/ShopifySection';
export { PaginateBlock } from './components/Paginate';
export { ForBlock } from './components/For';
export { ThemeProvider } from './components/ThemeProvider';
export { FormBlock } from './components/Form';
export { WrapperTag } from './components/WrapperTag';
export { SectionRef } from './components/SectionRef';
export { LiquidBlock } from './components/LiquidBlock';
export { StaticContent } from './components/StaticContent';

