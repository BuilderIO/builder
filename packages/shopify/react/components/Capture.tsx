import { BuilderStore, onChange, withBuilder } from '@builder.io/react';
import * as React from 'react';
import { CaptureBlockProps } from '../interfaces/component-props';

export class CaptureBlock extends React.Component<CaptureBlockProps> {
  ran = false;
  constructor(props: CaptureBlockProps) {
    super(props);
  }

  run() {
    const { expression, variableName, builderState } = this.props;

    if (expression && builderState?.context?.shopify) {
      onChange.target(builderState.state)[variableName] =
        builderState.context.shopify.liquid.render(expression, builderState.state);

      this.ran = true;
      return true;
    }

    return false;
  }

  render() {
    this.run();
    return null;
  }
}

withBuilder(CaptureBlock, {
  name: 'Shopify:Capture',
  hideFromInsertMenu: true,
  noWrap: true,
  inputs: [
    {
      name: 'expression',
      type: 'javascript',
    },
  ],
});
