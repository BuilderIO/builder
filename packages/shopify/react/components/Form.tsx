import { withBuilder, BuilderStoreContext, BuilderBlockComponent } from '@builder.io/react';
import * as React from 'react';
import { FormBlockProps } from '../interfaces/component-props';

export function FormBlock(props: FormBlockProps) {
  const { type, builderState, parameter, customAttributes, builderBlock, showErrors } = props;
  let form: any = {};
  let formAttributes: any = {
    method: 'post',
    acceptCharset: 'UTF-8',
  };

  // This looks for a global variable that we have server rendered by Shopify.
  // The goal is to get the Shopify form object and allow anything that needs data from
  // that form object to get it from the Builder state. https://shopify.dev/docs/themes/liquid/reference/objects/form
  if (window && (window as any).BuilderShopifyData && builderBlock?.id) {
    const formData = (window as any).BuilderShopifyData[builderBlock.id]?.form;

    if (formData) {
      Object.assign(form, formData);
    }
  }

  // Specific logic for each type of form tag, from https://shopify.dev/docs/themes/liquid/reference/tags/theme-tags#form
  if (type === 'activate_customer_password') {
    formAttributes.action = '/account/activate';
  } else if (type === 'contact') {
    formAttributes.action = '/contact#contact_form';
    formAttributes.id = 'contact_form';
    formAttributes.className = 'contact-form';
  } else if (type === 'currency') {
    formAttributes.action = '/cart/update';
    formAttributes.id = 'currency_form';
    formAttributes.encType = 'multipart/form-data';
    formAttributes.className = 'shopify-currency-form';
  } else if (type === 'customer') {
    formAttributes.action = '/contact#contact_form';
    formAttributes.id = 'contact_form';
    formAttributes.className = 'contact-form';
  } else if (type === 'create_customer') {
    formAttributes.action = '/account';
    formAttributes.id = 'create_customer';
  } else if (type === 'customer_address') {
    formAttributes.action = '/account/addresses';
    formAttributes.id = 'address_form_new';
  } else if (type === 'customer_login') {
    formAttributes.action = '/account/login';
  } else if (type === 'guest_login') {
    formAttributes.action = '/account/login';
    formAttributes.id = 'customer_login_guest';
  } else if (type === 'product') {
    const product = builderState?.context.shopify.liquid.get(parameter, builderState.state);
    formAttributes.action = '/cart/add';
    formAttributes.id = `product_form_${product?.id}`;
    formAttributes.className = 'shopify-product-form';
    formAttributes.encType = 'multipart/form-data';
  } else if (type === 'recover_customer_password') {
    formAttributes.action = '/account/recover';
  } else if (type === 'reset_customer_password') {
    formAttributes.action = '/account/reset';
  } else if (type === 'storefront_password') {
    formAttributes.action = '/password';
    formAttributes.id = 'login_form';
    formAttributes.className = 'storefront-password-form';
  } else if (type === 'new_comment') {
    formAttributes.action = '/blogs/news/my-blog-post/comments#comment_form';
    formAttributes.id = 'comment_form';
    formAttributes.className = 'comment-form';
  }

  if (customAttributes?.length) {
    customAttributes.forEach(attribute => {
      const [key, value] = attribute.split(':');
      let useValue = value?.trim();

      if (useValue && builderState?.state) {
        useValue = builderState.context.shopify.liquid.render(
          `{{${useValue}}}`,
          builderState.state
        );
      }

      if (key === 'class' && formAttributes.className && useValue) {
        formAttributes.className = `${formAttributes.className} ${useValue?.trim() || ''}`;
      } else if (useValue) {
        formAttributes[key] = useValue;
      } else if (key && !useValue) {
        // This is an attribute that does not have a value but should be added to the attribute list (e.g. boolean attribute)
        formAttributes[key] = true;
      }
    });
  }

  return (
    <BuilderStoreContext.Consumer>
      {store => (
        <BuilderStoreContext.Provider
          value={{
            ...store,
            state: {
              ...store.state,
              form,
            },
          }}
        >
          <form {...formAttributes}>
            <input type="hidden" name="form_type" value={type} />
            <input type="hidden" name="utf8" value="âœ“" />
            {type === 'currency' && (
              <input
                type="hidden"
                name="return_to"
                value={builderState?.state?.location?.pathname}
              />
            )}
            {showErrors && form?.errors && <span>{form.errors}</span>}
            {type === 'guest_login' && <input type="hidden" name="guest" value="true" />}
            {props.builderBlock?.children?.map(item => (
              <BuilderBlockComponent block={item} key={item.id} />
            ))}
          </form>
        </BuilderStoreContext.Provider>
      )}
    </BuilderStoreContext.Consumer>
  );
}

withBuilder(FormBlock, {
  name: 'Shopify:Form',
  hideFromInsertMenu: true,
  noWrap: true,
  inputs: [
    {
      name: 'type',
      type: 'string',
    },
    {
      name: 'parameter',
      type: 'string',
    },
    {
      name: 'customAttributes',
      type: 'array',
    },
  ],
});
