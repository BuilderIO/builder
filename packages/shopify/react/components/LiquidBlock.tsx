import React, { useState, useEffect, useRef } from 'react';
import { Builder, BuilderElement, BuilderStore, builder } from '@builder.io/react';
import { findAndRunScripts } from '../functions/find-and-run-scripts';
import { serializeLiquidArgs } from '../functions/serialize-liquid-args';

interface LiquidBlockProps {
  options?: Record<string, number | boolean | string>;
  templatePath?: string;
  builderBlock?: BuilderElement;
  builderState?: BuilderStore;
}
const cache: { [key: string]: string } = {};

const refs =
  Builder.isBrowser &&
  Array.from(document.querySelectorAll('[builder-liquid-block]')).reduce((memo, item) => {
    memo[item.getAttribute('builder-liquid-block')!] = item;
    return memo;
  }, {} as { [key: string]: Element });

export const LiquidBlock = ({
  options,
  builderBlock,
  templatePath,
  builderState,
}: LiquidBlockProps) => {
  const [html, setHtml] = useState<string | null>(null);
  const ref = useRef<HTMLDivElement | null>(null);
  const blockName = templatePath?.split('/')[1].replace(/\.liquid$/, '')!;

  useEffect(() => {
    const blockId = builderBlock?.id;
    const node = blockId && refs && refs[blockId];
    const args = serializeLiquidArgs(options);
    const cacheKey = blockName + args;

    if (!node || Builder.isEditing || Builder.isPreviewing) {
      if (cache[cacheKey]) {
        return;
      }

      // Convert `sections/foo.liquid` -> `foo`
      if (!blockName) {
        return;
      }

      if (cache[cacheKey]) {
        setHtml(cache[cacheKey]);
        return;
      }

      const previewThemeID = Builder.isBrowser && window.Shopify?.theme?.id || builderState?.state?.location?.query?.preview_theme_id;
      fetch(
        `${builder.host}/api/v1/shopify/data/render-liquid-snippet?snippet=${blockName}&apiKey=${
          builderState?.context.apiKey
        }&args=${encodeURIComponent(args)}${
          previewThemeID ? `&preview_theme_id=${previewThemeID}` : ''
        }`
      )
        .then(res => res.json())
        .then(json => {
          const text = json.liquidSnippet;
          cache[cacheKey] = text;
          setHtml(text);
          setTimeout(() => {
            if (ref.current) {
              findAndRunScripts(ref.current);
            }
          }, 10);
        });
    } else {
      if (node && ref.current) {
        ref.current.parentNode?.replaceChild(node, ref.current);
      }
    }
  }, [options, blockName]);

  return (
    <div
      ref={ref}
      builder-liquid-block={builderBlock?.id}
      className="builder-liquid-block"
      dangerouslySetInnerHTML={{
        __html: html || cache[blockName + serializeLiquidArgs(options)] || '',
      }}
    />
  );
};

Builder.registerComponent(LiquidBlock, {
  name: 'Shopify:LiquidBlock',
  hideFromInsertMenu: true,
  inputs: [
    {
      name: 'templatePath',
      type: 'string',
      advanced: true,
    },
    {
      name: 'options',
      type: 'LiquidBlockOptions',
    },
  ],
});
