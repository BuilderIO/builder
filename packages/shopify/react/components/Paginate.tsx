import {
  BuilderStore,
  withBuilder,
  BuilderStoreContext,
  BuilderBlockComponent,
  BuilderElement,
} from '@builder.io/react';
import * as React from 'react';

interface AssignBlockProps {
  expression?: string;
  limit?: string;
  builderState?: BuilderStore;
  builderBlock?: BuilderElement;
}

export function PaginateBlock(props: AssignBlockProps) {
  const defaultPaginationLimit = 15;
  const { expression, builderState, limit } = props;
  let paginate: any = {};

  if (
    expression &&
    builderState?.context?.shopify &&
    builderState.state.collection?.products?.length
  ) {
    const totalProducts = builderState.state.collection.products_count;
    const itemLimit =
      builderState.context.shopify.liquid.get(limit, builderState.state) || defaultPaginationLimit;

    let currentPage = 1;
    if (builderState.state.location.query.page) {
      // this is undefined, should probably add support to query params on location
      currentPage = builderState.state.location.query.page;
    } else if (window) {
      const urlParams = new URLSearchParams(window.location.search);
      currentPage = parseInt(urlParams.get('page') || '1', 10);
    }

    paginate = {
      current_page: currentPage,
      current_offset: currentPage * itemLimit,
      items: builderState.state.collection.products.length,
      parts: [],
      next: {},
      previous: {},
      page_size: itemLimit,
      pages: Math.ceil(totalProducts / itemLimit),
      expression: builderState.context.shopify.liquid.render(expression, builderState.state),
    };

    if (currentPage < paginate.pages) {
      paginate.next.url = `?page=${currentPage + 1}`;
    }

    if (currentPage > 1) {
      paginate.previous.url = `?page=${currentPage - 1}`;
    }
  }

  return (
    <BuilderStoreContext.Consumer>
      {store => (
        <BuilderStoreContext.Provider
          value={{
            ...store,
            state: {
              ...store.state,
              paginate,
            },
          }}
        >
          {props.builderBlock?.children?.map(item => (
            <BuilderBlockComponent block={item} key={item.id} />
          ))}
        </BuilderStoreContext.Provider>
      )}
    </BuilderStoreContext.Consumer>
  );
}

withBuilder(PaginateBlock, {
  name: 'Shopify:Paginate',
  hideFromInsertMenu: true,
  noWrap: true,
  inputs: [
    {
      name: 'expression',
      type: 'javascript',
    },
  ],
});
