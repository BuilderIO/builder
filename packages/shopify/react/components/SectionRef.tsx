import React, { useState, useEffect, useRef } from 'react';
import { Builder, BuilderElement } from '@builder.io/react';
import { findAndRunScripts } from '../functions/find-and-run-scripts';

interface SectionRefProps {
  class?: string;
  section?: string;
  builderBlock?: BuilderElement;
}
const cache: { [key: string]: string } = {};

const refs =
  Builder.isBrowser &&
  Array.from(document.querySelectorAll('[builder-shopify-section]')).reduce((memo, item) => {
    memo[item.getAttribute('builder-shopify-section')!] = item;
    return memo;
  }, {} as { [key: string]: Element });

export const SectionRef = (props: SectionRefProps) => {
  const [html, setHtml] = useState<string | null>(null);
  const sectionName = props.section?.split('/')[1].replace(/\.liquid$/, '')!;

  const ref = useRef<HTMLDivElement | null>(null);

  useEffect(() => {
    const blockId = props.builderBlock?.id;
    const node = blockId && refs && refs[blockId];
    if (!node) {
      if (cache[sectionName]) {
        return;
      }
      // Convert `sections/foo.liquid` -> `foo`
      if (!sectionName) {
        return;
      }
      if (cache[sectionName]) {
        setHtml(cache[sectionName]);
        return;
      }
      // This may look strange, but it is how Shopify's section rendering API
      // works https://shopify.dev/docs/themes/sections/section-rendering-api
      // TODO: also send ?preview_theme_id and save that as a cookie or wrap this API
      // specifically with /shopify/v1/data etc
      fetch(
        `https://cdn.builder.io/api/v1/proxy-api?url=${encodeURIComponent(
          `https://${location.host +
            location.pathname.replace(
              '/apps/builder/preview',
              ''
            )}?section_id=${sectionName}&${location.search.replace('?', '')}`
        )}`
      )
        .then(res => res.text())
        .then(text => {
          cache[sectionName] = text;
          setHtml(text);
          setTimeout(() => {
            if (ref.current) {
              findAndRunScripts(ref.current);
            }
          }, 10);
        });
    } else {
      if (node && ref.current) {
        ref.current.replaceWith(node);
      }
    }
  }, [sectionName]);

  return (
    <div
      ref={ref}
      builder-shopify-section={props.builderBlock?.id}
      className="builder-shopify-section"
      dangerouslySetInnerHTML={{ __html: html || cache[sectionName] || '' }}
    />
  );
};

Builder.registerComponent(SectionRef, {
  name: 'Shopify:SectionRef',
  hideFromInsertMenu: true,
});
