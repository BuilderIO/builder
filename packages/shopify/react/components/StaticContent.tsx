import React, { useEffect, useRef } from 'react';
import { BuilderElement, Builder } from '@builder.io/sdk';
import { BuilderBlocks, withBuilder } from '@builder.io/react';

const refs: Record<string, Element> = {};

if (Builder.isBrowser) {
  try {
    Array.from(document.querySelectorAll('.builder-static-content')).forEach(el => {
      const id = (el as HTMLDivElement).dataset.builderStaticId;
      if (id) {
        refs[id] = el;
      }
    });
  } catch (err) {
    console.error('Builder replace nodes error:', err);
  }
}

interface StaticContentProps {
  builderBlock?: BuilderElement;
}

const StaticContentComponent: React.SFC<StaticContentProps> = props => {
  const ref = useRef<HTMLDivElement | null>(null);

  useEffect(() => {
    if (Builder.isEditing || Builder.isPreviewing) {
      return;
    }

    const blockId = props.builderBlock?.id;
    if (blockId && refs[blockId] && ref.current) {
      ref.current.parentNode?.replaceChild(refs[blockId], ref.current);
    }
  }, []);

  return (
    <div className="builder-static-content" data-builder-static-id={props.builderBlock?.id} ref={ref}>
      {(Builder.isEditing || Builder.isPreviewing ||  Builder.isServer) &&
        props.builderBlock &&
        <BuilderBlocks child parent={props.builderBlock} blocks={props.builderBlock.children} />}
    </div>
  );
};

export const StaticContent = withBuilder(StaticContentComponent, {
  name: 'Shopify:StaticContent',
  static: true,
  image:
    'https://cdn.builder.io/api/v1/image/assets%2FYJIGb4i01jvw0SRdL5Bt%2Fc312e4bf11b64af0b43958aab099b927',
  canHaveChildren: true,
  defaultStyles: {
    // height: '200px',
    // how to disable styling
    marginTop: '0px'
  },
  defaultChildren: [
    {
      '@type': '@builder.io/sdk:Element',
      responsiveStyles: {
        large: {
          textAlign: 'center',
        },
      },
      component: {
        name: 'Text',
        options: {
          text: '<p>wrap content that should be static</p>',
        },
      },
    },
  ],
});
