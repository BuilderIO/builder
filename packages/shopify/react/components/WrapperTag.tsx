import * as React from 'react';
import {
  withBuilder,
  BuilderStore,
  BuilderElement,
  BuilderBlockComponent,
} from '@builder.io/react';

interface ConditionalTag extends Omit<BuilderElement, 'children'> {
  meta: { renderIf: string; }
}

interface WrapperTagProps {
  builderBlock: BuilderElement;
  conditionalTags: ConditionalTag[];
  builderState: BuilderStore;
}

function fastClone<T>(obj: T): T {
  return JSON.parse(JSON.stringify(obj));
}

/**
 * WrapperTag represent a set of tags that wrap its children conditionally
 */
export const WrapperTag: React.FC<WrapperTagProps> = ({
  conditionalTags,
  builderBlock,
  builderState,
}) => {
  if (builderState.context?.shopify) {
    const liquid = builderState.context.shopify.liquid;
    const validTags = conditionalTags.filter(
      tag => liquid.render(tag.meta.renderIf, builderState.state) === 'true'
    );
    const tags = fastClone(validTags) as BuilderElement[];

    if (tags.length === 0) {
      return <BuilderBlockComponent block={builderBlock} />;
    }

    const head = tags[0];
    let node = head;
    let i = 0;
    while (node) {
      i++;
      if (tags[i]) {
        node.children = [tags[i]];
      }
      // A temporary tag, only needed for debugging reasons, doesn't play a role in rendering
      delete node.component;
      node = tags[i];
    }
    tags[tags.length - 1].children = builderBlock.children;

    return <BuilderBlockComponent block={head} />;
  }
  return <></>;
};

withBuilder(WrapperTag, {
  name: 'Shopify:WrapperTag',
  canHaveChildren: true,
  noWrap: true,
  hideFromInsertMenu: true,
});
