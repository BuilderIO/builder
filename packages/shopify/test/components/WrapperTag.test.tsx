import * as React from 'react';
import { WrapperTag } from '../../react/components/WrapperTag';
import { mockStateWithShopify, text, builderComponentIdRegex, el } from '../modules/helpers';
import * as reactTestRenderer from 'react-test-renderer';

const conditionalTags = [
  {
    '@type': '@builder.io/sdk:Element',
    bindings: {},
    id: 'random1',
    component: {
      name: 'Shopify:ConditionalTag',
      options: {
        hash: '98e7248a0def13d04f49695de1adda440178bfce',
        name: 'opencondtag',
        tag: 'a',
      },
    },
    meta: {
      branchIndex: 0,
      originalIndex: 0,
      renderIf: '{% if thing %}true{% endif %}',
    },
    noWrap: true,
    properties: {},
    tagName: 'a',
  },
  {
    '@type': '@builder.io/sdk:Element',
    id: 'random2',
    bindings: {},
    component: {
      name: 'Shopify:ConditionalTag',
      options: {
        hash: '98e7248a0def13d04f49695de1adda440178bfce',
        name: 'opencondtag',
        tag: 'span',
      },
    },
    meta: {
      branchIndex: 1,
      originalIndex: 0,
      renderIf: '{% if thing %} {% else %}true{% endif %}',
    },
    noWrap: true,
    properties: {
      class: 'inelese',
    },
    tagName: 'span',
  },
] as any;

describe('Condition', () => {
  it('renders snapshot correctly', () => {
    const truthy = JSON.stringify(
      reactTestRenderer
        .create(
          <WrapperTag
            builderBlock={el({ children: [text('Should be a under a link')] })}
            conditionalTags={conditionalTags}
            builderState={mockStateWithShopify({ thing: true })}
          />
        )
        .toJSON()
    ).replace(builderComponentIdRegex, '');

    const falsy = JSON.stringify(
      reactTestRenderer
        .create(
          <WrapperTag
            builderBlock={el({ children: [text('Should be under a  span')] })}
            conditionalTags={conditionalTags}
            builderState={mockStateWithShopify({ thing: false })}
          />
        )
        .toJSON()
    ).replace(builderComponentIdRegex, '');

    expect({ truthy, falsy }).toMatchSnapshot();
  });
});
