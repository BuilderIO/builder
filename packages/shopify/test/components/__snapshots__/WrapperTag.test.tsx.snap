// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Condition renders snapshot correctly 1`] = `
Object {
  "falsy": "{\\"type\\":\\"span\\",\\"props\\":{\\"style\\":{},\\"className\\":\\"builder-block builder-random2 inelese css-0\\",\\"builder-id\\":\\"builder-random2\\"},\\"children\\":[{\\"type\\":\\"div\\",\\"props\\":{\\"style\\":{},\\"className\\":\\"builder-block  builder-has-component css-0\\",\\"builder-id\\":\\"\\"},\\"children\\":[{\\"type\\":\\"span\\",\\"props\\":{\\"className\\":\\"builder-text css-1qggkls\\",\\"dangerouslySetInnerHTML\\":{\\"__html\\":\\"Should be under a  span\\"}},\\"children\\":null}]}]}",
  "truthy": "{\\"type\\":\\"a\\",\\"props\\":{\\"style\\":{},\\"className\\":\\"builder-block builder-random1 css-0\\",\\"builder-id\\":\\"builder-random1\\"},\\"children\\":[{\\"type\\":\\"div\\",\\"props\\":{\\"style\\":{},\\"className\\":\\"builder-block  builder-has-component css-0\\",\\"builder-id\\":\\"\\"},\\"children\\":[{\\"type\\":\\"span\\",\\"props\\":{\\"className\\":\\"builder-text css-1qggkls\\",\\"dangerouslySetInnerHTML\\":{\\"__html\\":\\"Should be a under a link\\"}},\\"children\\":null}]}]}",
}
`;
